plonky2/src/plonk/circuit_builder.rs

```diff
@@ -173,44 +172,13 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
         (0..n).map(|_i| self.add_virtual_target()).collect()
     }
 
-    pub fn add_virtual_target_arr<const N: usize>(&mut self) -> [Target; N] {
-        [0; N].map(|_| self.add_virtual_target())
-    }
-
-    pub fn add_virtual_hash(&mut self) -> HashOutTarget {
-        HashOutTarget::from_vec(self.add_virtual_targets(4))
-    }
-
-    pub fn add_virtual_cap(&mut self, cap_height: usize) -> MerkleCapTarget {
-        MerkleCapTarget(self.add_virtual_hashes(1 << cap_height))
-    }
-
-    pub fn add_virtual_hashes(&mut self, n: usize) -> Vec<HashOutTarget> {
-        (0..n).map(|_i| self.add_virtual_hash()).collect()
-    }
-
-    pub(crate) fn add_virtual_merkle_proof(&mut self, len: usize) -> MerkleProofTarget {
-        MerkleProofTarget {
-            siblings: self.add_virtual_hashes(len),
-        }
-    }
-
-    pub fn add_virtual_extension_target(&mut self) -> ExtensionTarget<D> {
-        ExtensionTarget(self.add_virtual_targets(D).try_into().unwrap())
-    }
-
-    pub fn add_virtual_extension_targets(&mut self, n: usize) -> Vec<ExtensionTarget<D>> {
-        (0..n)
-            .map(|_i| self.add_virtual_extension_target())
-            .collect()
-    }
-
-    pub(crate) fn add_virtual_poly_coeff_ext(
+    pub fn add_virtual<'a, Config: Copy, A: FromTargets<'a, F, D, Config = Config>>(
         &mut self,
-        num_coeffs: usize,
-    ) -> PolynomialCoeffsExtTarget<D> {
-        let coeffs = self.add_virtual_extension_targets(num_coeffs);
-        PolynomialCoeffsExtTarget(coeffs)
+        c: Config,
+    ) -> A {
+        let len = A::len(c);
+        let mut targets = self.add_virtual_targets(len).into_iter();
+        A::from_targets(&mut targets, c)
```

Comment:
If we do go this direction, should probably assert `targets` is empty, so any error in `len()` is detected